### PR TITLE
Update react bootstrap table definitions to version 4.1.1

### PR DESCRIPTION
**Pull request by dawnmist**
The current template for react-bootstrap-table on DefinitelyTyped is for somewhere between version 2.4 - 2.8 (there were some options as far back as 2.4 that were missing). The current release for react-bootstrap-table is 4.0.6.

This updates the definitions to cover the new functionality that was missing, and to update stuff that had been altered/removed.

Tests have been created from example code within the react-bootstrap-table repository itself, as their documentation is also somewhat behind current functionality (several items are documented solely in their bug reports + the examples code that was put together for them). Using their example code for the tests means that the definitions should match the current interface for the library.

It's the first time I've submitted anything to DT, so I've done my best to follow the advice given. :)

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/AllenFang/react-bootstrap-table/blob/master/CHANGELOG.md
- [-] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by dawnmist** [comment]
Fixes #18840 and #10350.

**Comment by dawnmist** [comment]
@flaub, @alelode, @UJosue10 - any chance I can ask someone for a review so that this update for react-bootstrap-table types definitions can get merged?

**Review by josueus — APPROVED** [review]

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [inline types/react-bootstrap-table/index.d.ts]: This just becomes `any[]`. It does not seem to be an improvement over `string[] | number[]`.

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [inline types/react-bootstrap-table/index.d.ts]: This just becomes `any[]`.

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [inline types/react-bootstrap-table/index.d.ts]: This just becomes `any[]`, is that the intent?

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [inline types/react-bootstrap-table/index.d.ts]: `Row<any>` just becomes `any`.

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [inline types/react-bootstrap-table/index.d.ts]: This is only ever used below as `RowKey<any>`, so the constraint is effectively lost and `any` is used for those types instead ([example](https://www.typescriptlang.org/play/#src=type%20RowKey%3CT%20extends%20string%20%7C%20number%3E%20%3D%20T%3B%0D%0Aconst%20x%3A%20RowKey%3Cany%3E%20%3D%20true%3B)). If you want to restrict row keys to `string | number`, you need to define the type parameter and constraint on the various types or declarations below that use `RowKey`.

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [inline types/react-bootstrap-table/index.d.ts]: As with `RowKey` above, this is only ever used as `Cell<any>` and as such is effectively just `any`.

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [inline types/react-bootstrap-table/index.d.ts]: As with `RowKey` above, this is only ever used as `Row<any>` and as such is effectively just `any`.

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [reply types/react-bootstrap-table/index.d.ts]: As a suggestion, consider changing the definition of `BootstrapTableProps` to something like:

```ts
export interface BootstrapTableProps<TRow extends object = any> extends Props<BootstrapTable> {
  // ...
  data: TRow[];
  // ...
}
```

The same thing goes for `FooterData`, `SelectRow`, etc.

This change wouldn't be required for PR approval, but may align more closely with your original intent.

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [reply types/react-bootstrap-table/index.d.ts]: Suggestion (not required for approval): If you make `BootstrapTableProps` generic as I mention above, the definition for this would be:

```ts
trClassName?: string | (rowData: TRow, rowIndex: number) => string);
```

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [reply types/react-bootstrap-table/index.d.ts]: Suggestion (not required for approval): If you make `SelectRow` generic (e.g. `interface SelectRow<TRow extends object = any>`), you could be more typesafe here by specifying `selected` as: `selected?: Array<keyof TRow> | number[]`.

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [inline types/react-bootstrap-table/index.d.ts]: Suggestion (not required for approval): If you make `CellEdit` generic over a given row (e.g. `interface CellEdit<TRow extends object = any>`), you could make `beforeSaveCell` more typesafe by defining it as:

```ts
  beforeSaveCell?<K extends keyof TRow>(row: TRow, cellName: K, cellValue: TRow[K], done: (isValid: boolean) => void): boolean | 1;
```

**Review by rbuckton — CHANGES_REQUESTED** [review]
The `RowKey`, `Row`, and `Cell` type aliases introduced in this change do not seem useful. I recommend they be removed. 

I have provided some additional suggestions that are not required for approval. In addition to the specific places where those suggestions are indicated, there are other numerous locations throughout the file where they could be applied.

**Comment by dawnmist** [comment]
@rbuckton Thank you very much for the feedback. I've removed the Row/Cell/RowKey definitions, and used generics for most of the interfaces where they could assist.

With the BootstrapTable and TableHeaderColumn ones, because those get defined as JSX tags I don't know of a way to be able to set the type for their interface generics in the code when using them. That meant they were always being treated as the default 'any' when using what you'd suggested, so using generics on those particular interfaces wasn't providing any extra type safety. I also looked at trying to put the types as generics on the functions that used them within those interfaces but that triggered the 'no useless generics' rules so again wasn't helpful.

I would *love* to be able to set the BootstrapTable so that whatever datatype was used for the 'data' array would flow into "TRow" for the BootstrapTableProps/BootstrapTable interface/class, and so that TRow combined with (and checked validity of spelling) the TableHeaderColumn 'dataField' as a TRow key would flow into the TableHeaderColumnProps/TableHeaderColumn interface. If you know of a way to get something like that to work, I'll be very glad to add it. Otherwise, I think this is ready for re-review.

**Comment by rbuckton** [comment]
Unfortunately we don't currently have a mechanism for specifying generics on JSX elements in that fashion.

File: types/react-bootstrap-table/index.d.ts
Comment (rbuckton) [inline types/react-bootstrap-table/index.d.ts]: Judging from the changes to the tests, since you cannot flow `TRow` from `BootstrapTable` through to `TableHeaderColumn` the generics seem like more of a hindrance than a value. Since there's still some value in the type safety provided by generics, I would recommend adding type parameter defaults for these, such as `export interface SelectRow<TRow extends object = any>`.

**Review by rbuckton — CHANGES_REQUESTED** [review]

**Comment by RyanCavanaugh** [comment]
@dawnmist Please address comments from the code reviewers.

File: types/react-bootstrap-table/index.d.ts
Comment (dawnmist) [reply types/react-bootstrap-table/index.d.ts]: I've had a look at doing that, but I'm concerned about the result.

Taking the SelectRow as the example, by adding the "= any" you can define it without the type parameter. But if you do define it without the type parameter, the typescript type checker quite merrily accepts completely different items for the TRow parameter for the function parameters in SelectRow, and doesn't tell the user that there may be a problem.

e.g. The following passes typechecking with `export interface SelectRow<TRow extends object = any>`:
```
const selectRow: SelectRow {
  mode: 'checkbox',
  bgColor:
   (row: Job, isSelect: boolean) =>
      (isSelect ? (row.id < 2 ? 'blue' : row.id < 4 ? 'red' : 'yellow') : null),
  onSelectAll:
    (isSelected: boolean, rows: Product[]) =>
      (isSelected ? products.map((row) => row.id) : [])
}
```
Adding the type parameter to SelectRow as either `SelectRow<Product>` or `SelectRow<Job>` results in the type checker correctly identifying that TRow shouldn't be of both types and showing an error. Without the parameter (so that it uses the default "= any") it appears to completely disable the type checking for the generic.

That's why I didn't put the generics on the interfaces that were used to define the fields themselves in the JSX tags, as I found that the "default any" was completely disabling any type safety based on the generic.

Is that really the intent (i.e. to make typechecking opt-in only, without telling the user that they may want to opt-in)? Are there any interfaces in particular where defining the type while creating the object is problematical enough to want to silently disable typechecking for that interface as a default setting?

Comment (dawnmist) [reply types/react-bootstrap-table/index.d.ts]: @rbuckton Not sure if I needed to list your name for you to see my questions in response to your review comments.

**Comment by dawnmist** [comment]
@RyanCavanaugh @rbuckton From what I have tested, the code review comments appear to be asking me to make the default settings for the types to be 'typechecking = disabled'. I have tried asking for further detail as to why that setting would be necessary and received no response. I am concerned about the types being published in a form that only gives an illusion of type-checking (which is what the review comment appears to be asking me to do), and wondering what I'm missing.

Could I please get an explanation as to what the "=any" on the generic types is intended to achieve, and why a default value of disabling typechecking is necessary? I don't understand why this is desirable.

**Comment by dawnmist** [comment]
@rbuckton I've made the requested changes, and added a single updated definition to cover a change to the react-bootstrap-table library that was released recently.

**Comment by RyanCavanaugh** [comment]
Approved by a listed owner. PR ready to merge pending express review by a maintainer.

**Comment by RyanCavanaugh** [comment]
@rbuckton - Thanks for your review of this PR! Can you please look at the new code and update your review status if appropriate?

**Comment by DanielRosenwasser** [comment]
Thanks for the PR!

**Pull request merged**
